tests/testswitch.nim

```diff
@@ -602,6 +571,105 @@ suite "Switch":
 
     waitFor(testSwitch())
 
+  test "e2e should allow dropping peer from connection events":
+    proc testSwitch() {.async, gcsafe.} =
+      var awaiters: seq[Future[void]]
+
+      let rng = newRng()
+      # use same private keys to emulate two connection from same peer
+      let peerInfo = PeerInfo.init(PrivateKey.random(rng[]).tryGet())
+
+      var switches: seq[Switch]
+      var done = newFuture[void]()
+      var onConnect: Future[void]
+      proc hook(peerId: PeerID, event: ConnEvent) {.async, gcsafe.} =
+        case event.kind:
+        of ConnEventKind.Connected:
+          await onConnect
+          await switches[0].disconnect(peerInfo.peerId) # trigger disconnect
+        of ConnEventKind.Disconnected:
+          check not switches[0].isConnected(peerInfo.peerId)
+          await sleepAsync(1.millis)
```

Comment:
This is needed, otherwise the complete is synchronous and doesn't give the even't loop time to process anything else.